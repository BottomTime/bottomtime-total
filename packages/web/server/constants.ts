import { ApiClientOptions } from '@bottomtime/api';

import { StateTree } from 'pinia';

export type PageOptions = {
  appTitle: string;
  pageTitle: string;
  head: string;
  content: string;
  initialState: string;
};

export type RenderResult = {
  head?: string;
  html: string;
  initialState: string;
};

export type RenderFunc = (
  url: string,
  initialState: Record<string, StateTree>,
  clientOptions: ApiClientOptions,
) => Promise<RenderResult>;
