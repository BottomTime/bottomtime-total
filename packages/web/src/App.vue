<template>
  <div v-if="isLoading"></div>
  <div v-else>
    <SnackBar />
    <NavBar />
    <RouterView />
    <PageFooter />
    <CookieWarning />
  </div>
</template>

<script lang="ts" setup>
import { onBeforeMount, ref } from 'vue';

import CookieWarning from './components/main/CookieWarning.vue';
import { Dispatch, useStore } from '@/store';
import { inject } from '@/helpers';
import NavBar from '@/components/main/NavBar.vue';
import PageFooter from '@/components/main/PageFooter.vue';
import SnackBar from '@/components/main/SnackBar.vue';
import { ApiClientKey } from '@/injection-keys';

const client = inject(ApiClientKey);
const store = useStore();
const isLoading = ref(true);

onBeforeMount(async () => {
  try {
    const currentUser = await client.users.getCurrentUser();
    await store.dispatch(Dispatch.SignInUser, currentUser);
  } catch (error) {
    console.error(error);
  } finally {
    isLoading.value = false;
  }
});
</script>

<style>
@import './assets/bulmaswatch.min.css';
</style>
