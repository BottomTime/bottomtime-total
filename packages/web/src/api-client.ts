import { ApiClient } from '@bottomtime/api';

import { InjectionKey, inject } from 'vue';

export const ApiClientKey: InjectionKey<ApiClient> = Symbol('ApiClient');
export function useClient(): ApiClient {
  const client = inject(ApiClientKey);

  if (!client) {
    throw new Error(
      'Api Client has not been initialized. Did you forget to call `app.provide(key, client)`?',
    );
  }

  return client;
}
