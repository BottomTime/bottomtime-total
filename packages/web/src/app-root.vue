<template>
  <SnackBar />
  <div
    id="modal"
    class="font-content text-grey-900 bg-blue-200 dark:text-grey-100 dark:bg-blue-950"
  ></div>
  <div id="drawer"></div>

  <div
    class="md:container mx-auto font-content text-grey-900 bg-blue-200 dark:text-grey-100 dark:bg-blue-950 rounded-b-xl shadow-md shadow-white dark:shadow-grey-700 opacity-95"
  >
    <!-- Placeholder element. Modal dialogs will be "teleported" here to get the z-indexing right. -->

    <section class="mt-16 p-4">
      <RouterView></RouterView>
    </section>
    <PageFooter />
  </div>
  <NavBar />
</template>

<script setup lang="ts">
import { onBeforeMount } from 'vue';

import { ToastType } from './common';
import NavBar from './components/core/nav-bar.vue';
import PageFooter from './components/core/page-footer.vue';
import SnackBar from './components/core/snack-bar.vue';
import { Config } from './config';
import { useErrors, useToasts } from './store';

const toasts = useToasts();
const errors = useErrors();

onBeforeMount(() => {
  if (!Config.isSSR) {
    if (errors.renderError) {
      if (!Config.isProduction) {
        /* eslint-disable-next-line no-console */
        console.error(errors.renderError);
      }
      toasts.toast({
        id: 'server-prerender-error',
        message:
          'A server error has occurred and the page may not have loaded correctly. We apologize for the inconvenience. Please try again later as we investigate the issue.',
        type: ToastType.Error,
      });
    }
  }
});
</script>
