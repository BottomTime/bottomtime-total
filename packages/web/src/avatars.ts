import { AvatarSize, ListAvatarURLsResponseDTO } from '@bottomtime/api';

export type AvatarURLs = ListAvatarURLsResponseDTO['sizes'];

export function getAvatarURL(
  baseUrl: string | undefined,
  size: AvatarSize,
): string | undefined {
  if (!baseUrl) return undefined;

  if (/^\/api\/users\/[a-z0-9]+([_.-][a-z0-9]+)*\/avatar/i.test(baseUrl)) {
    return baseUrl.endsWith('/') ? `${baseUrl}${size}` : `${baseUrl}/${size}`;
  }

  return baseUrl;
}
