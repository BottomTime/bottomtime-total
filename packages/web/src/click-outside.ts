import { Directive, DirectiveBinding } from 'vue';

type HTMLElementWithClickOutside = HTMLElement & {
  clickOutsideEvent: (e: MouseEvent) => void;
};

export const clickOutside: Directive = {
  beforeMount: (el: HTMLElementWithClickOutside, binding: DirectiveBinding) => {
    el.clickOutsideEvent = (event) => {
      // here I check that click was outside the el and his children
      if (!(el == event.target || el.contains(event.target as Node))) {
        // and if it did, call method provided in attribute value
        binding.value();
      }
    };
    document.addEventListener('click', el.clickOutsideEvent);
  },
  unmounted: (el: HTMLElementWithClickOutside) => {
    document.removeEventListener('click', el.clickOutsideEvent);
  },
};
