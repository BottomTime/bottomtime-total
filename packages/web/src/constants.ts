import { z } from 'zod';

export enum ProfileVisibility {
  Private = 'private',
  FriendsOnly = 'friends',
  Public = 'public',
}

// UI Elements
export interface DropDownOption {
  value: string;
  text?: string;
  disabled?: boolean;
}

export const ProfileVisibilityOptions: Readonly<DropDownOption[]> = [
  {
    text: 'Everyone',
    value: ProfileVisibility.Public,
  },
  {
    text: 'Only My Dive Buddies',
    value: ProfileVisibility.FriendsOnly,
  },
  {
    text: 'Just Me',
    value: ProfileVisibility.Private,
  },
];

export enum TextBoxSize {
  Small = 'is-small',
  Normal = 'is-normal',
  Large = 'is-large',
}

export const MessageBoxStyle = {
  Danger: 'is-danger',
  Dark: 'is-dark',
  Info: 'is-info',
  Link: 'is-link',
  Success: 'is-success',
  Primary: 'is-primary',
  Warning: 'is-warning',
} as const;

// Depth
export enum DepthUnit {
  Meters = 'm',
  Feet = 'ft',
}

export const DepthSchema = z.object({
  depth: z.number(),
  unit: z.nativeEnum(DepthUnit),
});
export type Depth = z.infer<typeof DepthSchema>;

export const DepthUnitOptions: DropDownOption[] = [
  { text: 'Meters', value: 'm' },
  { text: 'Feet', value: 'ft' },
];

// Pressure
export enum PressureUnit {
  Bar = 'bar',
  PSI = 'psi',
}

export const PressureUnitOptions: DropDownOption[] = [
  { text: 'Bar', value: 'bar' },
  { text: 'PSI', value: 'psi' },
];

// Temperature
export enum TemperatureUnit {
  Celsius = 'C',
  Fahrenheit = 'F',
}

export const TemperatureUnitOptions: DropDownOption[] = [
  { text: '° Celsius', value: 'C' },
  { text: '° Fahrenheit', value: 'F' },
];

// Weight
export enum WeightUnit {
  Kilograms = 'kg',
  Pounds = 'lbs',
}

export const WeightUnitOptions: DropDownOption[] = [
  { text: 'Kilograms', value: 'kg' },
  { text: 'Pounds', value: 'lbs' },
];

// GPS Coordinates
export const GpsCoordinatesSchema = z.object({
  lat: z.number(),
  lon: z.number(),
});
export type GpsCoordinates = z.infer<typeof GpsCoordinatesSchema>;

// Misc.

// Email regex was copied from Zod repository to match Zod validation...
// https://github.com/colinhacks/zod/blob/master/src/types.ts#L568C3-L568C71
export const EmailRegex =
  /^([A-Z0-9_+-]+\.?)*[A-Z0-9_+-]@([A-Z0-9][A-Z0-9-]*\.)+[A-Z]{2,}$/i;

export const PasswordStrengthRegex =
  /^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[~`!@#$%^&*()-_+=}{}[\]<>,./?|\\/]).{8,50}$/;

export const UsernameRegex = /^[a-z0-9]+([_.-][a-z0-9]+)*$/i;

export enum UserRole {
  User = 'user',
  Admin = 'admin',
}
