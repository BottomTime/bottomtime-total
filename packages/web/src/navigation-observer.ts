import { InjectionKey, inject } from 'vue';
import { RouteLocationNormalized, Router } from 'vue-router';

import { useLogger } from './logger';

const log = useLogger('NavigationObserver');

export type CanNavigateGuard = (
  to: RouteLocationNormalized,
  from: RouteLocationNormalized,
) => boolean | Promise<boolean>;

export interface INavigationObserver {
  subscribe: (guard: CanNavigateGuard) => void;
  unsubscribe: (guard: CanNavigateGuard) => void;
}

export const NavigationObserverKey: InjectionKey<INavigationObserver> =
  Symbol('NavigationObserver');

export class NavigationObserver implements INavigationObserver {
  private readonly guards: CanNavigateGuard[] = [];

  private async canNavigate(
    this: NavigationObserver,
    to: RouteLocationNormalized,
    from: RouteLocationNormalized,
    next: (canNavigate: boolean) => void,
  ): Promise<void> {
    if (this.guards.length === 0) {
      next(true);
      return;
    }

    log.debug(
      'Attempting to navigate from',
      from.fullPath,
      'to',
      to.fullPath,
      'using',
      this.guards.length,
      'navigation guards...',
    );
    let canNavigate = true;

    for (const guard of this.guards) {
      canNavigate = await guard(to, from);
      if (!canNavigate) break;
    }

    log.debug('Navigation guards passed.');
    next(canNavigate);
  }

  constructor(router: Router) {
    router.beforeEach(this.canNavigate.bind(this));
  }

  subscribe(guard: CanNavigateGuard) {
    log.debug('Subscribing to navigation guard');
    this.guards.push(guard);
  }

  unsubscribe(guard: CanNavigateGuard) {
    const index = this.guards.indexOf(guard);
    if (index > -1) this.guards.splice(index, 1);
  }
}

export function useNavigationObserver(): INavigationObserver {
  const observer = inject(NavigationObserverKey);

  if (!observer) {
    throw new Error(
      'NavigationObserver has not been provided. Did you remember to call `app.provide()`?',
    );
  }

  return observer;
}
