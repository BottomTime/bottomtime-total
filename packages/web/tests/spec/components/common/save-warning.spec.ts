import {
  ComponentMountingOptions,
  flushPromises,
  mount,
} from '@vue/test-utils';

import SaveWarning from 'src/components/common/save-warning.vue';
import { NavigationObserverKey } from 'src/navigation-observer';
import { ConfirmDialog } from 'tests/constants';
import { createRouter } from 'tests/fixtures/create-router';
import { MockNavigationObserver } from 'tests/mock-navigation-observer';
import { Router } from 'vue-router';

const SaveWarningElement = '[data-testid="save-warning"]';
const SaveButton = '#btn-save-warning-save';

describe('SaveWarning component', () => {
  let router: Router;
  let navigationObserver: MockNavigationObserver;
  let opts: ComponentMountingOptions<typeof SaveWarning>;

  beforeEach(async () => {
    router = createRouter([
      {
        path: '/from',
        component: SaveWarning,
      },
      {
        path: '/to',
        component: SaveWarning,
      },
    ]);
    await router.push('/from');

    navigationObserver = new MockNavigationObserver(router);
    opts = {
      props: {
        isDirty: true,
      },
      global: {
        plugins: [router],
        provide: {
          [NavigationObserverKey as symbol]: navigationObserver,
        },
        stubs: {
          teleport: true,
        },
      },
    };
  });

  it('will render if dirty flag is set', () => {
    const wrapper = mount(SaveWarning, opts);
    const warning = wrapper.get(SaveWarningElement);
    expect(warning.isVisible()).toBe(true);
    expect(warning.html()).toMatchSnapshot();
  });

  it('will be hidden if dirty flag is unset', async () => {
    const wrapper = mount(SaveWarning, opts);
    await wrapper.setProps({ isDirty: false });
    expect(wrapper.find(SaveWarningElement).exists()).toBe(false);
  });

  it('will emit a "save" event if save button is clicked', async () => {
    const wrapper = mount(SaveWarning, opts);
    await wrapper.get(SaveButton).trigger('click');
    expect(wrapper.emitted('save')).toHaveLength(1);
  });

  it('will render with "save" button disabled if "isSaving" flag is set', async () => {
    const wrapper = mount(SaveWarning, opts);
    await wrapper.setProps({ isSaving: true });
    expect(wrapper.get(SaveButton).attributes('disabled')).toBeDefined();
  });

  it('will prevent a navigation and show a confirmation dialog', async () => {
    const wrapper = mount(SaveWarning, opts);
    await expect(router.push('/to')).resolves.toBeInstanceOf(Error);
    expect(wrapper.get(ConfirmDialog.Confirm).isVisible()).toBe(true);
  });

  it('will allow a user to confirm a navigation', async () => {
    const wrapper = mount(SaveWarning, opts);
    await router.push('/to');
    await wrapper.get(ConfirmDialog.Confirm).trigger('click');
    await flushPromises();
    expect(router.currentRoute.value.fullPath).toBe('/to');
  });

  it('will allow a user to cancel a navigation', async () => {
    const wrapper = mount(SaveWarning, opts);
    await router.push('/to');
    await wrapper.get(ConfirmDialog.Cancel).trigger('click');
    await flushPromises();
    expect(router.currentRoute.value.fullPath).toBe('/from');
  });
});
