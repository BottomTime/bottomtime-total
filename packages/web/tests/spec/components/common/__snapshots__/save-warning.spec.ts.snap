// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveWarning component will render if dirty flag is set 1`] = `
"<div class="w-full fixed left-0 bottom-0 font-bold text-lg" data-testid="save-warning">
  <div class="bg-warn-dark container mx-auto rounded-t-lg px-3 py-1 flex justify-between items-baseline shadow-md shadow-grey-800/60">
    <p>Your changes have not yet been saved!</p><button id="btn-save-warning-save" class="text-grey-950 p-2 m-0 text-sm border-1 border-grey-800 shadow-sm shadow-grey-800 rounded-md disabled:text-grey-500 bg-secondary hover:bg-secondary-hover" data-testid="btn-save-warning-save" type="button">
      <!--v-if--> Save Changes
    </button>
  </div>
</div>"
`;
